### PR TITLE
Master fix

### PR DESCRIPTION
**Pull request by kingpinXD**

### CODE REVIEW

File: scripts/ethereum/main.go [hunk 1]
```diff
@@ -48,7 +46,7 @@ var (
 	TestTokenABI     = contract.ERC20BasicABI
 	LockRedeemERCABI = contract.LockRedeemERCABI
 	// LockRedeemERC20ABI = contract.ContextABI
-	LockRedeemContractAddr      = "0xd2629c4282E2946a6d762120164887e8f48C6AeB"
+	LockRedeemContractAddr      = "0xDc9F35C56344A769D4908044Ab35AE00Faa01f1A"
```
Comment (pav-kap) [inline scripts/ethereum/main.go:49]: why did u change the smart contract address?

File: scripts/resetDev [hunk 2]
```diff
@@ -7,9 +7,9 @@ rm -rf $OLDATA/devnet/*
 olfullnode init devnet --dir $OLDATA/devnet --nonvalidators 1 \
 --empty_blocks \
 --loglevel 4 \
---eth_rpc HTTP://127.0.0.1:7545 \
-#--total_funds 1000000000 \
-#--deploy_smart_contracts \
+--eth_rpc https://rinkeby.infura.io/v3 \
```
Comment (pav-kap) [inline scripts/resetDev:10]: Are these changes  needed?

File: cmd/olfullnode/devnet.go [hunk 3]
```diff
@@ -600,7 +600,7 @@ func deployethcdcontract(conn string, nodeList []node) (*ethchain.ChainDriverOpt
 	lock_period := big.NewInt(25)
 
 	tokenSupplyTestToken := new(big.Int)
-	validatorInitialFund := big.NewInt(300000000000000000) //300000000000000000
+	validatorInitialFund := big.NewInt(0) //300000000000000000
```
Comment (pav-kap) [inline cmd/olfullnode/devnet.go:603]: Is this needed?

File: event/eth_sign_redeem.go [hunk 4]
```diff
@@ -121,21 +121,30 @@ func (j *JobETHSignRedeem) DoMyJob(ctx interface{}) {
...
+		BroadcastReportFinalityETHTx(ctx.(*JobsContext), j.TrackerName, j.JobID, false)
+	}
+	//Signature confirmed
 	if success {
 		ethCtx.Logger.Debug("validator Sign Confirmed | validator Address (SIGNER):", ethCtx.GetValidatorETHAddress().Hex(), "| User Eth Address :", msg.From().Hex())
 		j.Status = jobs.Completed
 		return
 	}
+
 	if j.RetryCount >= 0 && !success {
```
Comment (pav-kap) [inline event/eth_sign_redeem.go:138]: No point in checking !success.. as it will always be true at this place. You can try to move this condition above , if that makes sense

Comment (pav-kap) [reply event/eth_sign_redeem.go:138]: Or probably check for err != nil

File: scripts/ethereum/main.go [hunk 5]
```diff
@@ -365,18 +363,18 @@ func redeem() []byte {
 		return nil
 	}
 
-	//acc := accReply.Accounts[0]
+	acc := accReply.Accounts[0]
```
Comment (pav-kap) [inline scripts/ethereum/main.go:366]: Are these changes related to this pr?

File: chains/ethereum/online_chain_driver.go [hunk 6]
```diff
@@ -136,7 +136,7 @@ func (acc *ETHChainDriver) SignRedeem(fromaddr common.Address, redeemAmount *big
 	if err != nil {
 		return nil, err
 	}
-	gasLimit := uint64(6721974)
+	gasLimit := uint64(700000)
```
Comment (pav-kap) [inline chains/ethereum/online_chain_driver.go:139]: Can we add this as a constant somewhere.. so that we dont need to change it at multiple places?

File: event/eth_sign_redeem.go [hunk 7]
```diff
@@ -87,20 +87,20 @@ func (j *JobETHSignRedeem) DoMyJob(ctx interface{}) {
 	tx, err := cd.DecodeTransaction(rawTx)
 	if err != nil {
 		ethCtx.Logger.Error("Error Decoding Bytes from RaxTX :", j.GetJobID(), err)
-		return
+		//return
```
Comment (pav-kap) [inline event/eth_sign_redeem.go:90]: This has to be implemented thoroughly.. I m not sure if its good idea to remove the return here. Which means, even though the decode fails, it will go to line 93 to get the transaction message. Is that a good idea?

If the oETH is already burnt in these cases, then we should refund in all these cases, instead of just returning here...

You also need to consider the situation of lock here...Not just redeem

File: event/eth_sign_redeem.go [hunk 8]
```diff
@@ -87,20 +87,20 @@ func (j *JobETHSignRedeem) DoMyJob(ctx interface{}) {
...
 	}
 
 	msg, err := cd.GetTransactionMessage(tx)
 	if err != nil {
 		ethCtx.Logger.Error("Error in decoding transaction as message : ", j.GetJobID(), err)
-		return
+		//return
```
Comment (pav-kap) [inline event/eth_sign_redeem.go:96]: same as above

File: event/eth_sign_redeem.go [hunk 9]
```diff
@@ -87,20 +87,20 @@ func (j *JobETHSignRedeem) DoMyJob(ctx interface{}) {
...
 	}
 
 	addr := ethCtx.GetValidatorETHAddress()
 	txReceipt, err := cd.VerifyReceipt(tx.Hash())
 	if err != nil {
 		ethCtx.Logger.Debug("Trying to confirm RedeemTX sent by User Receipt :", err)
-		return
+		//return
```
Comment (pav-kap) [inline event/eth_sign_redeem.go:103]: same as above

File: client/request_types.go [hunk 10]
```diff
@@ -10,6 +10,7 @@
 package client
```
Comment (pav-kap) [inline client/request_types.go:10]: Why does this commit have all these changes?
@kingpinXD are you sure your pr is right?

File: chains/ethereum/online_chain_driver.go (hunk 6 above)
Comment (ws4charlie) [reply chains/ethereum/online_chain_driver.go:139]: Constant?

File: service/query/service.go [hunk 11]
```diff
@@ -31,14 +32,15 @@ func Name() string {
 	return "query"
 }
 
-func NewService(ctx client.ExtServiceContext, balances *balance.Store, currencies *balance.CurrencySet, validators *identity.ValidatorStore,
+func NewService(ctx client.ExtServiceContext, balances *balance.Store, currencies *balance.CurrencySet, validators *identity.ValidatorStore, witnesses *identity.WitnessStore,
```
Comment (ws4charlie) [inline service/query/service.go:35]: Is this new listWitness service related to the fix of ETH Redeem?

File: scripts/resetDev (hunk 2 above)
Comment (ws4charlie) [reply scripts/resetDev:10]: same question as Pavan.

File: scripts/ethereum/main.go (hunk 5 above)
Comment (ws4charlie) [reply scripts/ethereum/main.go:366]: Are these changes only related to integration/unit tests?

File: scripts/resetDev (hunk 2 above)
Comment (ws4charlie) [reply scripts/resetDev:10]: Only for devnet/local test, right?

File: app/context.go [hunk 12]
```diff
@@ -209,6 +209,7 @@ func (ctx *context) Services() (service.Map, error) {
 		Cfg:          ctx.cfg,
 		NodeContext:  ctx.node,
 		ValidatorSet: identity.NewValidatorStore("v", storage.NewState(ctx.chainstate)),
+		WitnessSet:   identity.NewWitnessStore("w", storage.NewState(ctx.chainstate)),
```
Comment (ws4charlie) [inline app/context.go:212]: Is this related to the fix or not?

File: cmd/olfullnode/devnet.go (hunk 3 above)
Comment (ws4charlie) [reply cmd/olfullnode/devnet.go:603]: Should this PR be made from **MASTER** to the deployed taged version?
